### DISCUSSION TITLE
FormCollection - multiple default_renderer in holders/forms

### DISCUSSION THREAD

**Discussion by gumish** [jrief/django-formset#43] [Q&A]
Is there any way to use multiple 'default_renderer's in a FormCollection please?

In EditFormCollection(view) I would like to render the:
- main_object form - as a classic form
- child formset - as a table form

Example:
```
from formset.renderers.bootstrap import FormRenderer

class TeamCollection(FormCollection):
    min_siblings = 0
    team = TeamForm()
    default_renderer = FormRenderer(
        form_css_classes='row',
        field_css_classes='col my-1',
        label_css_classes='d-none'
    )

class CompanyCollection(FormCollection):
    default_renderer = FormRenderer()
    form = CompanyForm()  # main_object = classic form
    teams = TeamCollection()  # formset = table
```
But default_renderer works only for the parent CompanyCollection. The TeamCollection's default_renderer is ignored.

**Comment by jrief** [comment]
Isn't that documented? Apparently not good enough.

You can add a renderer other than the default for sub-collections.
Either pass it a parameter when instantiating your `TeamCollection`,

```python
class CompanyCollection(FormCollection):
    default_renderer = FormRenderer()
    form = CompanyForm()
    teams = TeamCollection(renderer=...)
``` 

or add it to the `TeamCollection` directly:

```python
class TeamCollection(FormCollection):
    min_siblings = 0
    team = TeamForm()
    renderer = FormRenderer(
        form_css_classes='row',
        field_css_classes='col my-1',
        label_css_classes='d-none'
    )
```

**Reply by gumish** [reply]
I haven't found any mention about plain 'renderer' in documentation (only about 'default_renderer').

Now I tried both solutions written above but **no one unfortunately worked** :-(
The whole collection looks like according to parent 'default_renderer'.

![image](https://user-images.githubusercontent.com/7414504/226104966-ad4f6f88-8bed-4184-98bd-7abd3b96c93f.png)

**Comment by jrief** [comment]
Hmm, thought this was working already. Apparently no unit tests yet.
Will check asap.
Thanks for reporting.

**Reply by jrief** [reply]
In the meantime, please check if #37 works for you, so that I can close it.

**Reply by jrief** [reply]
This actually works already. Please check here: https://django-formset.fly.dev/bootstrap/contact
 There the `PhoneNumberForm` (wich is a sub-form of `PhoneNumberCollection` which itself is a sub-collection of `ContactCollection`) has a special `renderer` using the CSS classes
```json
{
    "form_css_classes": "row",
    "field_css_classes": {
        "phone_number": "mb-2 col-8", "label": "mb-2 col-4", "*": "mb-2 col-12"
    }
}
```

By doing so, the number and label field is aligned on the same line.

It's a bit tricky to see this from the sample code. This is because I wanted to keep the examples CSS-framework independent. Please check https://github.com/jrief/django-formset/blob/94550e1431e14c7c59f78b84d97ff3148eb85874/testapp/views.py#L203 to see how it's done.

I will create an example where it's easier to understand.

**Reply by gumish** [reply]
I see! I had checked the 'traverse_holders' https://github.com/jrief/django-formset/blob/94550e1431e14c7c59f78b84d97ff3148eb85874/testapp/views.py#L203
and realized that **'renderer' is here set to 'Form class'**, not to 'FormCollection class' like I was doing whole time!

So I had to do:
```
class TeamCollection(FormCollection):
    min_siblings = 0
    team = TeamForm(
        renderer=FormRenderer(
            form_css_classes='row',
            field_css_classes='col my-1',
            label_css_classes='d-none'
        )
    )
```
and then also my case works as expected. But it will be more logical if it works as described in my first comment. To set 'default_rendeder' to FormCollection.
